cpp/unreal_plugins/UrdfBot/Source/UrdfParser.cpp

```diff
@@ -0,0 +1,415 @@
+            joint_desc.effort_ = FCString::Atof(*node->GetAttribute(TEXT("effort")));
+        } else if (tag.Equals(TEXT("mimic"))) {
+            ASSERT(!has_mimic_node);
+            has_mimic_node = true;
+            joint_desc.joint_ = TCHAR_TO_UTF8(*node->GetAttribute(TEXT("joint")));
+            joint_desc.multiplier_ = FCString::Atof(*node->GetAttribute(TEXT("multiplier")));
+            joint_desc.offset_ = FCString::Atof(*node->GetAttribute(TEXT("offset")));
+        } else if (tag.Equals(TEXT("safety_controller"))) {
+            ASSERT(!has_safety_controller_node);
+            has_safety_controller_node = true;
+            joint_desc.soft_lower_limit_ = FCString::Atof(*node->GetAttribute(TEXT("soft_lower_limit")));
+            joint_desc.soft_upper_limit_ = FCString::Atof(*node->GetAttribute(TEXT("soft_upper_limit")));
+            joint_desc.k_position_ = FCString::Atof(*node->GetAttribute(TEXT("k_position")));
+            joint_desc.k_velocity_ = FCString::Atof(*node->GetAttribute(TEXT("k_velocity")));
+        }
+    }
+
+    ASSERT(has_parent_node);
+    ASSERT(has_child_node);
+    // revolute and prismatic joints require limit
+    ASSERT(!(joint_desc.type_ == UrdfJointType::Revolute || joint_desc.type_ == UrdfJointType::Prismatic) || has_limit_node);
+
+    return joint_desc;
+}
+
+UrdfRobotDesc UrdfParser::parseRobotNode(FXmlNode* robot_node)
+{
+    UrdfRobotDesc robot_desc;
+
+    // parse all top-level URDF nodes into their own dictionaries
+    for (auto& child_node : robot_node->GetChildrenNodes()) {
+        const FString& tag = child_node->GetTag();
+
+        if (tag.Equals(TEXT("link"))) {
+            UrdfLinkDesc link_desc = parseLinkNode(child_node);
+            ASSERT(!robot_desc.link_descs_.count(link_desc.name_));
+            robot_desc.link_descs_[link_desc.name_] = std::move(link_desc);
+        } else if (tag.Equals(TEXT("joint"))) {
+            UrdfJointDesc joint_desc = parseJointNode(child_node);
+            ASSERT(!robot_desc.joint_descs_.count(joint_desc.name_));
+            robot_desc.joint_descs_[joint_desc.name_] = std::move(joint_desc);
+        } else if (tag.Equals(TEXT("material"))) {
+            UrdfMaterialDesc material_desc = parseMaterialNode(child_node);
+            ASSERT(material_desc.name_ != "" && !material_desc.is_reference_);
+            ASSERT(!robot_desc.material_descs_.count(material_desc.name_));
+            robot_desc.material_descs_[material_desc.name_] = std::move(material_desc);
+        }
+    }
+
+    // for a robot with N links, there must be N-1 joints.
+    ASSERT(robot_desc.joint_descs_.size() + 1 == robot_desc.link_descs_.size());
+
+    // build link tree based on joints
+    for (auto& joint_desc_pair : robot_desc.joint_descs_) {
+        UrdfJointDesc* joint_desc = &(joint_desc_pair.second);
+        ASSERT(joint_desc);
+
+        UrdfLinkDesc* parent_link_desc = &robot_desc.link_descs_.at(joint_desc->parent_);
```

Comment:
```
        UrdfLinkDesc* parent_link_desc = &(robot_desc.link_descs_.at(joint_desc->parent_));
        UrdfLinkDesc* child_link_desc = &(robot_desc.link_descs_.at(joint_desc->child_));
```

Extra parentheses for clarity and consistency with the rest of the code in this function.

Reply:
fixed